### ISSUE TITLE
Access javascript variables from Lua and vice versa?

### ISSUE THREAD

**Issue by Sheepolution** [Teoxoy/lua-in-js#3]
Is it possible to communicate variables between Lua and Javascript with this library?

**Comment by Teoxoy** [comment]
The `run` and `runfile` functions output the return value of the script.
As for accessing javascript variables from lua, it's not currently possible. You could just pass the value of the js variable trough the lua script though. i.e.
```js
const text = 'Hello world!'
luaEnv.run(`print('${text}')`)
```
What is your usecase for this?

**Comment by Sheepolution** [comment]
I'm porting [LÖVE](https://love2d.org/) to Javascript. LÖVE uses Lua, so it would be nice to have code for LÖVE work on browser as well. Similar projects already exist with different Lua VMs/transpilers but they're quite outdated.

So I have this class called Love, which has properties (modules) that are instances of classes, and those classes have functions. So what I want to be able to do is calling `love.graphics.newImage(url)` from Lua to create a new image.

**Comment by Teoxoy** [comment]
I think this should be doable by adding a new global table `love` then exposing the LÖVE functionality through it. `love` would just be another lib like `math`, `table` and so on. You can take a look in the `lib` folder and see how the other global libs are implemented.

Before you start, make sure you don't need any of the [missing functionality](https://github.com/Teoxoy/lua-in-js#missing-functionality) listed on the readme as I don't have plans to implement those yet, if at all.
Feel free to ask more questions regarding this.

I will provide a function to register a global package so that you don't have to fork the repo.

**Comment by Sheepolution** [comment]
Cool, thanks!
And yes it's fine that it misses that functionality.
Looking in the lib folder I think I understand how to do this. Would all the modules (love.graphics, love.audio, etc.) need to be tables as well?

**Comment by Teoxoy** [comment]
I exposed `loadLib` in e45eda2b71b67bacbfd4b110cf1168a7efe2e29f
More info [here](https://github.com/Teoxoy/lua-in-js#create-a-global-library)

**Comment by Teoxoy** [comment]
Yes, `graphics`, `audio`, etc. would be nested tables in the main `love` table.

**Comment by Teoxoy** [comment]
On a side note, if you want to run a lua script multiple times (via `luaEnv.run`), the lua code will be transpiled to JS on every call of `run`. This adds unnecessary overhead so I might change the API of `run` and `runfile` so that they return a function that you call as many times as you want without transpiling on every call.

Might look like this:
```js
// parse
const luaScript = luaEnv.parse(`<lua_script>`) // and luaEnv.parseFile(`<file_path>`)
// run
luaScript.run()
```

Thoughts?

**Comment by Sheepolution** [comment]
Yes, sounds good. And thank you so much! I will try it out as soon as I can.

**Comment by Sheepolution** [comment]
Hey so it's working great so far. Only problem I'm having is loading the files. Right now I use sync xhr, but this is becoming deprecated, so I would somehow need to know and preload all the files that are going to be required. Images will also need to be preloaded so that you can use `image:getWidth()`. Else it would return 0 until the image is loaded.

Now I can think of a few solutions to these. One would be to have the player run a script so that the web-version of their game will be built, or that they have to manually insert which files they want to have preloaded. Another solution would be using async/await. Except, if I understand async/await correctly, this would mean that all functions, including the ones that are converted from Lua, need to be async. What I really like about that solution is that the user would be able to drop their LÖVE code in a folder and that the lib will do all the rest. A big disadvantage might be that it's slower, but if this is insignificant I think it's worth it.

So since I think this is too much to ask of you I'd like to fork the project and try to make an async version myself. But before I do that I would like your opinion. Do you think it would be doable? Would it be a little or a lot of work? Any advice would be great. I woudn't be surprised if your answer is that it's not really doable or that it would require a complete rework of the parser. If that's the case I'll have to accept that and look into the build script. Thanks.

**Comment by Teoxoy** [comment]
I think the best solution would be to use [`Promise`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)s.
Short example:
```ts
function loadImage(imgURL, onSuccess, onError) {
    const url = coerceArgToString(imgURL, 'loadImage', 1)
    const onsuccess = coerceArgToFunction(onSuccess, 'loadImage', 2)
    const onerror = coerceArgToFunction(onError, 'loadImage', 3)

    fetch(url)
        .then(img => {
            // transform html image into LÖVE representation and call onsuccess
        })
        .catch(err => {
            // handle errors and call onerror
        })
}
```
You can also have a function in the API that would load all resources in parallel using `Promise.all()`

Only downside I see with this is using callbacks but if you handle them well I think it might not be that big of a problem.

I am not sure if I got the whole picture but I think this should be much better than converting all the code to async/await - not sure even if it work to convert evrything to async/await in the first place.

**Comment by Sheepolution** [comment]
Well the problem with this is that, if I understand correctly, Lua will continue with whatever it's doing while the promise is being resolved.

```lua
local image = love.graphics.newImage("player.png")
print(image:getWidth())
```
It will output 0 because the promise is still in the process of being resolved. Are you suggesting that in LÖVE you should then have a callback like
```lua
function image:onSuccess()
    print(image:getWidth())
end
```
The problem with that approach is that I would like code that is made for LÖVE to work with this LÖVE for browser lib, with as few changes as possible. This would require a lot of changes.

My idea with the async/await is that my code at the top would be converted to something like 

```js
image = await love.graphics.newImage("player.png");
await console.log(await image.getWidth());
```
So it awaits the promise of getting the image, and after that it prints the width. The `console.log` and `getImage` wouldn't have to be async but that would be the result of changing the parser.

**Comment by Teoxoy** [comment]
I forgot to mention, `fetch(url)` will return a blob for image resources.
Check out this [link ](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch)for more info on using `fetch`.
This is a sample from that link:
```js
try {
  const response = await fetch('flowers.jpg');
  if (!response.ok) {
    throw new Error('Network response was not ok.');
  }
  const myBlob = await response.blob();
  const objectURL = URL.createObjectURL(myBlob);
  myImage.src = objectURL;
} catch (error) {
  console.log('There has been a problem with your fetch operation: ', error.message);
}
```
This would be the Promise version:
```ts
fetch('flowers.jpg')
    .then(res => {
        if (res.ok) return res
        else throw new Error('Network response was not ok.')
    })
    .then(res => res.blob())
    .then(blob => {
        const objURL = URL.createObjectURL(blob)
        const img = new Image()
        img.src = objURL
        // call onsuccess
    })
    .catch(err => {
        console.log('There has been a problem with your fetch operation: ', err.message)
        // call onerror
    })
```

**Comment by Teoxoy** [comment]
How does the original LÖVE implementation load resources? Is it always sync?

**Comment by Sheepolution** [comment]
Yes, it's always sync (well unless you use a thread but most people don't use that so yes all resources you load are sync).

**Comment by Teoxoy** [comment]
Does LÖVE perform any optimizations? Are all assets preloaded at startup? Because I would imagine loading resources sync and also sequentially to be quite slow. This would be even worse in a browser env where it takes even longer to load assets.

**Comment by Sheepolution** [comment]
As far as I know LÖVE does this directly. And yeah you're right it would probably be slower but I'm not sure how noticeable it would be.

**Comment by Teoxoy** [comment]
I thought about this some more and I think the best way to achieve this is to modify the parser to look for all occurrences of `newImage` (or any other sync operation that loads assets) then generate js code that would preload all these assets using `Promise.all`.

**Comment by Teoxoy** [comment]
@Sheepolution any updates on this?

**Comment by Sheepolution** [comment]
@Teoxoy I tried making everything async but that didn't really work. I've decided that this is not gonna work out and moved on to other projects. Sorry for not letting you know. You can close this issue if you want.

**Comment by Teoxoy** [comment]
Thanks for letting me know!
If you decide to work on it again and have more questions feel free to reopen the issue.
